src/js/node/events.js

```diff
@@ -386,21 +388,20 @@ async function* on(emitter, event, options = {}) {
 
   for (const evName of options?.close || []) {
     emitter.on(evName, () => {
-      emitter.emit(event, undefined);
       done = true;
     });
   }
 
-  while (!done) {
+  while (!done || !unconsumedEvents.isEmpty()) {
     if (!unconsumedEvents.isEmpty()) {
       yield Promise.$resolve(unconsumedEvents.shift());
-    }
-    if (!unconsumedErrors.isEmpty()) {
+    } else if (!unconsumedErrors.isEmpty()) {
       yield Promise.$reject(unconsumedErrors.shift());
+    } else {
+      const { promise, reject, resolve } = $newPromiseCapability(Promise);
+      unconsumedPromises.push({ reject, resolve });
+      yield promise;
```

Comment:
And a second thing about this generator function: To notify the `for await...of` loop that the last item is reached, this item should not be `yield`ed but `return`ed. This is not the case here, yet the execution still stops after the last item. Is this an "ungraceful" exit, since the JS engine notices somehow that the last returned promise can never be resolved?

Again, not an expert on generator functions. 😉